### PR TITLE
More resiliency when collecting logs

### PR DESCRIPTION
**Pull request by superseb**
https://github.com/rancher/rancher/issues/16609

### CODE REVIEW

File: logs/logs.go [hunk 1]
```diff
@@ -102,7 +101,10 @@ func DoLogs(ctx *cli.Context) error {
 			return err
 		}
 		f.Sync()
+		buf.Reset()
 	}
+	deleteLogCollectors(client)
```
Comment (alena1108) [inline logs/logs.go]: is there a reason why we don't delete log collector when the error happens on lines 98 and 101?

File: logs/logs.go [hunk 2]
```diff
@@ -127,19 +129,19 @@ func deployLogCollectors(client *kubernetes.Clientset) error {
...
+		logrus.Infof("waiting for DaemonSet [%s] to be ready..", LogCollectorDSName)
+
 		ds, err := client.AppsV1().DaemonSets(logCollectorDS.Namespace).Get(logCollectorDS.Name, v1.GetOptions{})
 		if err != nil {
 			return err
 		}
-		if ds.Status.NumberAvailable != ds.Status.DesiredNumberScheduled {
-			time.Sleep(time.Second * 1)
-			continue
+		if ds.Status.DesiredNumberScheduled > 0 && ds.Status.DesiredNumberScheduled == ds.Status.NumberReady {
```
Comment (alena1108) [inline logs/logs.go]: is there a reason for  `ds.Status.DesiredNumberScheduled > 0` check? Would `ds.Status.NumberReady` ever be 0?

**Comment by alena1108** [comment]
@superseb Besides the code being more resilient, could you explain which line in the change fixes 16609?

File: logs/logs.go (hunk 2 above)
Comment (superseb) [reply logs/logs.go]: This is the flow:

```
INFO[0000] waiting for DaemonSet [log-collector] to be ready..
INFO[0000] ds.Status.DesiredNumberScheduled: 0
INFO[0000] ds.Status.NumberReady: 0
INFO[0000] ds.Status.NumberAvailable: 0
INFO[0001] waiting for DaemonSet [log-collector] to be ready..
INFO[0001] ds.Status.DesiredNumberScheduled: 3
INFO[0001] ds.Status.NumberReady: 0
INFO[0001] ds.Status.NumberAvailable: 0
INFO[0002] waiting for DaemonSet [log-collector] to be ready..
INFO[0002] ds.Status.DesiredNumberScheduled: 3
INFO[0002] ds.Status.NumberReady: 1
INFO[0002] ds.Status.NumberAvailable: 1
INFO[0003] waiting for DaemonSet [log-collector] to be ready..
INFO[0003] ds.Status.DesiredNumberScheduled: 3
INFO[0003] ds.Status.NumberReady: 1
INFO[0003] ds.Status.NumberAvailable: 1
INFO[0004] waiting for DaemonSet [log-collector] to be ready..
INFO[0004] ds.Status.DesiredNumberScheduled: 3
INFO[0004] ds.Status.NumberReady: 1
INFO[0004] ds.Status.NumberAvailable: 1
INFO[0005] waiting for DaemonSet [log-collector] to be ready..
INFO[0005] ds.Status.DesiredNumberScheduled: 3
INFO[0005] ds.Status.NumberReady: 2
INFO[0005] ds.Status.NumberAvailable: 2
INFO[0006] waiting for DaemonSet [log-collector] to be ready..
INFO[0006] ds.Status.DesiredNumberScheduled: 3
INFO[0006] ds.Status.NumberReady: 2
INFO[0006] ds.Status.NumberAvailable: 2
INFO[0007] waiting for DaemonSet [log-collector] to be ready..
INFO[0007] ds.Status.DesiredNumberScheduled: 3
INFO[0007] ds.Status.NumberReady: 3
INFO[0007] ds.Status.NumberAvailable: 3
```

File: logs/logs.go (hunk 1 above)
Comment (superseb) [reply logs/logs.go]: Yes, I moved it back to `defer deleteLogCollectors(client)` after creating it.

**Comment by superseb** [comment]
Two things:

- There was no wait for the logs collecting pods to be ready, so they might still be collecting logs when the logs are being retrieved. Added a readiness check to make sure pods are ready before proceeding.
- `bytes.Buffer` needs to be reset/truncated between uses: https://github.com/rancher/system-tools/pull/17/files#diff-eb8f9e6743ee7ab75f5dba1570e3c5b5R105

**Pull request merged**